### PR TITLE
Documented From* impls in alloc/string.rs

### PR DESCRIPTION
**Pull request by duckymirror**
For #51430.

### CODE REVIEW

File: library/alloc/src/string.rs [hunk 1]
```diff
@@ -1741,6 +1761,11 @@ impl FromIterator<Box<str>> for String {
 
 #[stable(feature = "herd_cows", since = "1.19.0")]
 impl<'a> FromIterator<Cow<'a, str>> for String {
+    /// Converts an [`IntoIterator<Item = Cow<'a, str>>`](IntoIterator) into a [`String`].
+    ///
+    /// If `iter` isn't empty, the data is appended to its first element,
+    /// which is cloned it if it isn't already owned, resulting in an allocation on the heap.
```
Comment (shepmaster) [inline library/alloc/src/string.rs:1767]:
```suggestion
    /// which is cloned if it isn't already owned, resulting in an allocation on the heap.
```

File: library/alloc/src/string.rs [hunk 2]
```diff
@@ -2150,6 +2175,11 @@ pub type ParseError = core::convert::Infallible;
 #[stable(feature = "rust1", since = "1.0.0")]
 impl FromStr for String {
     type Err = core::convert::Infallible;
+    /// Converts a `&str` into a `String`.
+    ///
+    /// The result is allocated on the heap.
+    ///
```
Comment (shepmaster) [inline library/alloc/src/string.rs:2181]:
```suggestion
```

**Review by shepmaster — COMMENTED** [review]
Basically my few comments apply to all of the changes. I don't see the benefit in documentation of the form:

```rust
/// Adds one to the `u8` called `a` and returns it
fn add_one(a: u8) -> u8 { a + 1 }
```

Documentation should describe the things that the code doesn't/can't tell us, such as nuances or warnings or usage guidance. 

On the other hand, putting implementation details into the documentation is dangerous because a decision has to be made about if those details are things we want to commit to forever. See [the blurb in `std::io`](https://doc.rust-lang.org/std/io/index.html#platform-specific-behavior) as an example about guarding against that.

File: library/alloc/src/string.rs [hunk 3]
```diff
@@ -1687,6 +1687,10 @@ impl Clone for String {
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl FromIterator<char> for String {
+    /// Converts an [`IntoIterator<Item = char>`](IntoIterator) into a [`String`].
```
[the remark this answers is not in the document]
Comment (duckymirror) [reply library/alloc/src/string.rs:1690]: I followed these mentoring instructions: https://github.com/rust-lang/rust/issues/51430#issuecomment-398322434
Many others did too, so there are other examples of this line. I don't know whether it's written this way in other places though.

**Comment by camelid** [comment]
@duckymirror Could you address Shepmaster's review comments?

**Comment by JohnCSimon** [comment]
Ping from triage: 
@duckymirror - Thanks for your contribution, but unfortunately I'll have to close this due to inactivity, please feel free to reopen when you ready to proceed.

File: library/alloc/src/string.rs [hunk 4]
```diff
@@ -1687,6 +1687,10 @@ impl Clone for String {
...
+    ///
+    /// This appends the data in `iter` to a new [`String`].
+    /// The data is allocated on the heap if `iter` isn't empty.
```
[the remark this answers is not in the document]
Comment (notriddle) [reply library/alloc/src/string.rs:1693]: Just as an FYI:

* [skade's example comment](https://github.com/rust-lang/rust/issues/51430#issuecomment-398322434) actually writes it that way.
* If you add the doc comment to the trait impl, then rustdoc will still add a "[+] show undocumented items", but if you put the doc comment on the fn impl, then it won't.

**Pull request closed without merging**